storage/src/main/java/io/delta/storage/internal/S3LogStoreUtil.java

```diff
@@ -0,0 +1,119 @@
+ * limitations under the License.
+ */
+
+package io.delta.storage.internal;
+
+import com.amazonaws.services.s3.model.ListObjectsV2Request;
+import org.apache.hadoop.fs.*;
+import org.apache.hadoop.fs.s3a.*;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.HashSet;
+
+import static org.apache.hadoop.fs.s3a.Constants.DEFAULT_MAX_PAGING_KEYS;
+import static org.apache.hadoop.fs.s3a.Constants.MAX_PAGING_KEYS;
+import static org.apache.hadoop.fs.s3a.S3AUtils.iteratorToStatuses;
+
+
+/**
+ * Static utility methods for the S3SingleDriverLogStore.
+ *
+ * Used to trick the class loader so we can use methods of org.apache.hadoop:hadoop-aws without needing to load this as
+ * a dependency for tests in core.
+ */
+public final class S3LogStoreUtil {
+    private S3LogStoreUtil() {}
+
+    private static PathFilter ACCEPT_ALL = new PathFilter() {
+        @Override
+        public boolean accept(Path file) {
+            return true;
+        }
+
+        @Override
+        public String toString() {
+            return "ACCEPT_ALL";
+        }
+    };
+
+    /**
+     * Uses the S3ListRequest.v2 interface with the startAfter parameter to only list files
+     * which are lexicographically greater than resolvedPath.
+     */
+    public static RemoteIterator<S3AFileStatus> s3ListFrom(FileSystem fs, Path resolvedPath, Path parentPath) throws IOException {
+        S3AFileSystem s3afs = (S3AFileSystem) fs;
+        int maxKeys = S3AUtils.intOption(s3afs.getConf(), MAX_PAGING_KEYS, DEFAULT_MAX_PAGING_KEYS, 1);
+        Listing listing = s3afs.getListing();
+        // List files lexicographically after resolvedPath inclusive within the same directory
+        return listing.createFileStatusListingIterator(resolvedPath,
+                S3ListRequest.v2(
+                        new ListObjectsV2Request()
+                                .withBucketName(s3afs.getBucket())
+                                .withMaxKeys(maxKeys)
+                                .withPrefix(s3afs.pathToKey(parentPath))
+                                .withStartAfter(keyBefore(s3afs.pathToKey(resolvedPath)))
+                ), ACCEPT_ALL,
+                new Listing.AcceptAllButSelfAndS3nDirs(parentPath)
+        );
+    }
+
+    /**
+     * Uses the S3ListRequest.v2 interface with the startAfter parameter to only list files
+     * which are lexicographically greater than resolvedPath.
+     *
+     * Wraps s3ListFrom in an array. Contained in this class to avoid contaminating other
+     * classes with dependencies on recent Hadoop versions.
+     *
+     * TODO: Remove this method when iterators are used everywhere.
+     */
+    public static FileStatus[] s3ListFromArray(FileSystem fs, Path resolvedPath, Path parentPath) throws IOException {
+        try {
+            return iteratorToStatuses(S3LogStoreUtil.s3ListFrom(fs, resolvedPath, parentPath), new HashSet<>());
+        } catch (ClassCastException | NoClassDefFoundError e) {
```

Comment:
Generally, we prefer to wrap the narrowest lines that could throw these exceptions. Any other place throwing ClassCastException or NoClassDefFoundError will be caused by other issues, and we should not catch that and provide a misleading error message.

In addition, `ClassCastException` and `NoClassDefFoundError` are two different issues:
- `ClassCastException`: the user uses a wrong fs. Can we just catch this and improve it as `UnsupportedOperationException`?
- `NoClassDefFoundError`: the user doesn't provide hadoop-aws. This is not related to this PR. We can open a separate PR to improve the error message for this case if needed.

Reply:
Thank you. Addressed in https://github.com/delta-io/delta/pull/1210/commits/958a9e3ec69ceaf3989c455fbe3efe719b4106b2